source/common/tcp_proxy/upstream.h

```diff
@@ -301,6 +301,8 @@ class CombinedUpstream : public GenericUpstream, public Envoy::Router::RouterFil
   bool readDisable(bool disable) override;
   void setConnPoolCallbacks(std::unique_ptr<HttpConnPool::Callbacks>&& callbacks) {
     conn_pool_callbacks_ = std::move(callbacks);
+    auto is_ssl = upstream_request_->streamInfo().upstreamInfo()->upstreamSslConnection();
```

Comment:
this is outside of meaning of `setConnPoolCallbacks`, would be better to split it, also would be great to merge the ssl check and header setting to one place.